### PR TITLE
[CPDLP-1244] Allow a DfE user to change the status of an NPQ application from the support page

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

- Ticket: https://dfedigital.atlassian.net/browse/CPDLP-1244

### Changes proposed in this pull request

* New "Change to pending" link added to participant drill down support page

### Guidance to review

* Login as `finance@example.com`
* Go to the participant drill down page
* Enter participant id with NPQ Application

### CODE REVIEW

File: spec/forms/finance/change_lead_provider_approval_status_form_spec.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        cohort:
+      )
+    end
+    it { is_expected.to validate_inclusion_of(:confirmation).in_array(%w[yes no]) }
+
+    describe ".save" do
+      context "valid params" do
+        it "should change status" do
+          expect(form.save).to be true
+          expect(npq_application.reload.lead_provider_approval_status).to eql("pending")
```
Comment (StupidCodeFactory) [inline spec/forms/finance/change_lead_provider_approval_status_form_spec.rb:28]: Would you mind using RSpec's syntactic sugar to change for boolean methods?:
```
expect(npq_application.reload).to be_pending
```

File: spec/forms/finance/change_lead_provider_approval_status_form_spec.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        end
+      end
+
+      context "invalid params" do
+        let(:params) { { npq_application:, confirmation: "" } }
+
+        it "should not change status" do
+          expect(form.save).to be false
+          expect(npq_application.reload.lead_provider_approval_status).to eql("accepted")
```
Comment (StupidCodeFactory) [inline spec/forms/finance/change_lead_provider_approval_status_form_spec.rb:37]:
```suggestion
          expect(npq_application.reload).to be_accepted
```

File: spec/forms/finance/change_lead_provider_approval_status_form_spec.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        cohort:
+      )
+    end
+    it { is_expected.to validate_inclusion_of(:confirmation).in_array(%w[yes no]) }
+
+    describe ".save" do
+      context "valid params" do
+        it "should change status" do
+          expect(form.save).to be true
+          expect(npq_application.reload.lead_provider_approval_status).to eql("pending")
```
Comment (StupidCodeFactory) [inline spec/forms/finance/change_lead_provider_approval_status_form_spec.rb:58]:
```suggestion
          expect(npq_application.reload).to be_pending
```

File: spec/forms/finance/change_lead_provider_approval_status_form_spec.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        end
+      end
+
+      context "invalid params" do
+        let(:params) { { npq_application:, confirmation: "" } }
+
+        it "should not change status" do
+          expect(form.save).to be false
+          expect(npq_application.reload.lead_provider_approval_status).to eql("rejected")
```
Comment (StupidCodeFactory) [inline spec/forms/finance/change_lead_provider_approval_status_form_spec.rb:67]:
```suggestion
          expect(npq_application.reload).to be_rejected
```

File: app/forms/finance/change_lead_provider_approval_status_form.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module Finance
+  class ChangeLeadProviderApprovalStatusForm
+    include ActiveModel::Model
+
+    attr_accessor :npq_application, :confirmation
```
[the remark this answers is not in the document]
Comment (mooktakim) [reply app/forms/finance/change_lead_provider_approval_status_form.rb:7]: You can't properly do inclusion validation on a boolean field. So its best to keep string here as the UI allows for yes or no option.

File: app/views/finance/change_lead_provider_approval_statuses/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        <strong>Lead provider approval status</strong>
+        <br/>
+        <%= @npq_application.lead_provider_approval_status %>
+      </p>
+    </div>
+
+    <%= form_for @change_lead_provider_approval_status_form, url: finance_npq_application_change_lead_provider_approval_status_path(@npq_application) do |f| %>
+      <%= f.govuk_error_summary %>
+      <%= f.govuk_radio_buttons_fieldset :training_status, legend: { text: "Change the status to pending?", tag: 'h1', size: 'm' } do %>
+        <%= f.govuk_radio_button :confirmation, "yes", label: { text: "Yes" } %>
```
Comment (peteryates) [inline app/views/finance/change_lead_provider_approval_statuses/new.html.erb:18]:
```suggestion
        <%= f.govuk_radio_button :confirmation, "yes", label: { text: "Yes" }, link_erorrs: true %>
```

File: app/controllers/finance/change_lead_provider_approval_statuses_controller.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    def set_npq_application
+      @npq_application = NPQApplication.find(params[:npq_application_id])
+    end
+
+    def change_lead_provider_approval_status_form_params
+      return {} unless params.key?(:finance_change_lead_provider_approval_status_form)
+
+      params.require(:finance_change_lead_provider_approval_status_form).permit(
+        :confirmation,
```
Comment (peteryates) [inline app/controllers/finance/change_lead_provider_approval_statuses_controller.rb:32]: The name of the attribute here doesn't tell me what's being confirmed. Would it be clearer if it was something like `:change_status_to_pending`?

**Review by peteryates — COMMENTED** [review]
This looks good to me.

Left one comment regarding `link_errors` being present on the first item in a radio button fieldset and a question on naming.

File: app/services/npq/reject.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    def call
+      if npq_application.rejected?
+        npq_application.errors.add(:lead_provider_approval_status, :has_already_been_rejected)
+        return false
+      end
+
+      if npq_application.accepted?
+        npq_application.errors.add(:lead_provider_approval_status, :invalid, message: "Once accepted an application cannot change state")
```
Comment (StupidCodeFactory) [inline app/services/npq/reject.rb:24]: We put all the error messages in the localisation files

File: app/views/finance/change_lead_provider_approval_statuses/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        <strong>Lead provider approval status</strong>
+        <br/>
+        <%= @npq_application.lead_provider_approval_status %>
+      </p>
+    </div>
+
+    <%= form_for @change_lead_provider_approval_status_form, url: finance_npq_application_change_lead_provider_approval_status_path(@npq_application) do |f| %>
+      <%= f.govuk_error_summary %>
+      <%= f.govuk_radio_buttons_fieldset :training_status, legend: { text: "Change the status to pending?", tag: 'h1', size: 'm' } do %>
+        <%= f.govuk_radio_button :change_status_to_pending, "yes", label: { text: "Yes" }, link_erorrs: true %>
```
Comment (peteryates) [inline app/views/finance/change_lead_provider_approval_statuses/new.html.erb:18]: @mooktakim  sorry there was a typo in my previous suggestion 😭 

```suggestion
        <%= f.govuk_radio_button :change_status_to_pending, "yes", label: { text: "Yes" }, link_errors: true %>
```

**Review by leoapost — APPROVED** [review]

**Review by peteryates — APPROVED** [review]
This looks good to me

**Pull request merged**